zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs

```diff
@@ -34,23 +43,37 @@ pub(crate) struct ErrorOOGCallGadget<F> {
 impl<F: Field> ExecutionGadget<F> for ErrorOOGCallGadget<F> {
     const NAME: &'static str = "ErrorOutOfGasCall";
 
-    const EXECUTION_STATE: ExecutionState = ExecutionState::ErrorOutOfGasCALL;
+    const EXECUTION_STATE: ExecutionState = ExecutionState::ErrorOutOfGasCall;
 
     fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
         let opcode = cb.query_cell();
         cb.opcode_lookup(opcode.expr(), 1.expr());
-        // TODO: add CallCode etc. when handle ErrorOutOfGasCALLCODE in furture
-        // implementation
+        let is_call = IsZeroGadget::construct(cb, opcode.expr() - OpcodeId::CALL.expr());
+        let is_callcode = IsZeroGadget::construct(cb, opcode.expr() - OpcodeId::CALLCODE.expr());
+        let is_delegatecall =
+            IsZeroGadget::construct(cb, opcode.expr() - OpcodeId::DELEGATECALL.expr());
+        let is_staticcall =
+            IsZeroGadget::construct(cb, opcode.expr() - OpcodeId::STATICCALL.expr());
+
+        // We do the responsible opcode check explicitly here because we're not
+        // using the SameContextGadget for CALL, CALLCODE, DELEGATECALL or
+        // STATICCALL.
         cb.require_equal(
-            "ErrorOutOfGasCall opcode is Call",
-            opcode.expr(),
-            OpcodeId::CALL.expr(),
+            "ErrorOutOfGasCall opcode should be CALL, CALLCODE, DELEGATECALL or STATICCALL",
+            is_call.expr() + is_callcode.expr() + is_delegatecall.expr() + is_staticcall.expr(),
+            1.expr(),
         );
 
         let rw_counter_end_of_reversion = cb.query_cell();
         let tx_id = cb.call_context(None, CallContextFieldTag::TxId);
         let is_static = cb.call_context(None, CallContextFieldTag::IsStatic);
```

Comment:
This seems not used, perhaps we could remove it. (gas calculation doesn't depends on if it's read-only context).

Reply:
I try to reference the code of `CallOpGadget` and find a constraint of `has_value` and `is_static` as below. I copied it into `ErrorOOGCallGadget` in 8622af9.
```
cb.condition(call_gadget.has_value.expr(), |cb| {
    cb.require_zero(
        "CALL with value must not be in static call stack",
        is_static.expr(),
    );
});
```